src/NuGet.Core/NuGet.Packaging/Signing/Utility/CertificateChainUtility.cs

```diff
@@ -0,0 +1,167 @@
+            }
+
+            policy.ExtraStore.AddRange(additionalCertificates);
+
+            policy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+            policy.RevocationMode = X509RevocationMode.Online;
+
+            if (certificateType != NuGetVerificationCertificateType.Timestamp)
+            {
+                policy.VerificationTime = verificationTime;
+            }
+        }
+
+
+        internal static bool BuildCertificateChain(X509Chain chain, X509Certificate2 certificate, out X509ChainStatus[] status)
+        {
+            if (certificate == null)
+            {
+                throw new ArgumentNullException(nameof(certificate));
+            }
+
+            var buildSuccess = chain.Build(certificate);
+            status = new X509ChainStatus[chain.ChainStatus.Length];
+            chain.ChainStatus.CopyTo(status, 0);
+
+            // Check if time is not in the future
+            return buildSuccess && !CertificateUtility.IsCertificateValidityPeriodInTheFuture(certificate);
+        }
+
+
+        // Ignore some chain status flags to special case them
+        internal const X509ChainStatusFlags NotIgnoredCertificateFlags =
+            // To set the not ignored flags we have to turn on all the flags and then manually turn off ignored flags
+            (~(X509ChainStatusFlags)0) &                      // Start with all flags
+                                                              // These flags are ignored because they are known special cases
+            (~X509ChainStatusFlags.NotTimeValid) &
+            (~X509ChainStatusFlags.NotTimeNested) &           // Deprecated and therefore ignored.
+            (~X509ChainStatusFlags.Revoked) &
+            (~X509ChainStatusFlags.RevocationStatusUnknown) &
+            (~X509ChainStatusFlags.CtlNotTimeValid) &
+            (~X509ChainStatusFlags.OfflineRevocation);
+
+        internal static bool ChainStatusListIncludesStatus(X509ChainStatus[] chainStatuses, X509ChainStatusFlags status, out IEnumerable<X509ChainStatus> chainStatus)
+        {
+            chainStatus = chainStatuses
+                .Where(x => (x.Status & status) != 0);
+
+            if (chainStatus.Any())
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        internal static bool TryGetStatusMessage(X509ChainStatus[] chainStatuses, X509ChainStatusFlags status, out IEnumerable<string> messages)
+        {
+            messages = null;
+
+            if (ChainStatusListIncludesStatus(chainStatuses, status, out var chainStatus))
+            {
+                messages = GetMessagesFromChainStatuses(chainStatus.ToArray());
+
+                return true;
+            }
+
+            return false;
+        }
+
+        internal static IEnumerable<string> GetMessagesFromChainStatuses(X509ChainStatus[] chainStatuses)
+        {
+            return chainStatuses
+                .Select(x => x.StatusInformation?.Trim())
```

Comment:
Sad to see [`X509ChainStatusFlags.ToString()`](https://github.com/NuGet/NuGet.Client/pull/1896/files#diff-da57037b885995340b10db626a834dd4L105) go.  It was a language-agnostic framework error code.

Reply:
Even though `X509ChainStatusFlags.ToString()` was language-agnostic I believe `StatusInformation` provides a better user experience and a more readable error message.